sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/WatchForKafkaTopicPartitions.java

```diff
@@ -0,0 +1,184 @@
+  private static final String COUNTER_NAMESPACE = "watch_kafka_topic_partition";
+
+  private final Duration checkDuration;
+  private final SerializableFunction<Map<String, Object>, Consumer<byte[], byte[]>>
+      kafkaConsumerFactoryFn;
+  private final Map<String, Object> kafkaConsumerConfig;
+  private final @Nullable SerializableFunction<TopicPartition, Boolean> checkStopReadingFn;
+  private final List<String> topics;
+  private final @Nullable Instant startReadTime;
+  private final @Nullable Instant stopReadTime;
+
+  public WatchForKafkaTopicPartitions(
+      @Nullable Duration checkDuration,
+      SerializableFunction<Map<String, Object>, Consumer<byte[], byte[]>> kafkaConsumerFactoryFn,
+      Map<String, Object> kafkaConsumerConfig,
+      @Nullable SerializableFunction<TopicPartition, Boolean> checkStopReadingFn,
+      List<String> topics,
+      @Nullable Instant startReadTime,
+      @Nullable Instant stopReadTime) {
+    this.checkDuration = firstNonNull(checkDuration, DEFAULT_CHECK_DURATION);
+    this.kafkaConsumerFactoryFn = kafkaConsumerFactoryFn;
+    this.kafkaConsumerConfig = kafkaConsumerConfig;
+    this.checkStopReadingFn = checkStopReadingFn;
+    this.topics = topics;
+    this.startReadTime = startReadTime;
+    this.stopReadTime = stopReadTime;
+  }
+
+  @Override
+  public PCollection<KafkaSourceDescriptor> expand(PBegin input) {
+    return input
+        .apply(Impulse.create())
+        .apply(
+            "Match new TopicPartitions",
+            Watch.growthOf(
+                    new WatchPartitionFn(kafkaConsumerFactoryFn, kafkaConsumerConfig, topics))
+                .withPollInterval(checkDuration))
+        .apply(ParDo.of(new ConvertToDescriptor(checkStopReadingFn, startReadTime, stopReadTime)));
+  }
+
+  private static class ConvertToDescriptor
+      extends DoFn<KV<byte[], TopicPartition>, KafkaSourceDescriptor> {
+
+    private final @Nullable SerializableFunction<TopicPartition, Boolean> checkStopReadingFn;
+    private final @Nullable Instant startReadTime;
+    private final @Nullable Instant stopReadTime;
+
+    private ConvertToDescriptor(
+        @Nullable SerializableFunction<TopicPartition, Boolean> checkStopReadingFn,
+        @Nullable Instant startReadTime,
+        @Nullable Instant stopReadTime) {
+      this.checkStopReadingFn = checkStopReadingFn;
+      this.startReadTime = startReadTime;
+      this.stopReadTime = stopReadTime;
+    }
+
+    @ProcessElement
+    public void processElement(
+        @Element KV<byte[], TopicPartition> partition,
+        OutputReceiver<KafkaSourceDescriptor> receiver) {
+      TopicPartition topicPartition = Objects.requireNonNull(partition.getValue());
+      if (checkStopReadingFn == null || !checkStopReadingFn.apply(topicPartition)) {
+        Counter foundedTopicPartition =
+            Metrics.counter(COUNTER_NAMESPACE, topicPartition.toString());
+        foundedTopicPartition.inc();
```

Comment:
fyi - runner metrics systems are not designed to support 'big data'. Dataflow only guarantees a few thousands of metrics being supported (which I guess is enough in this case?) so just worth keeping in mind.